src/r1cs/circuits/jolt_duplicated.circom

```diff
@@ -0,0 +1,316 @@
+pragma circom 2.1.6;
+
+/*  Constants: written as functions because circom.
+*/
+function N_CHUNKS() {return 6;}
+function L_CHUNK() {return 11;}
+function L_MS_CHUNK() {return 9;}
+function N_FLAGS() {return 15;}
+
+function WIT_PER_STEP() {return 6 * 3 + 1 + 2 * 3 + N_FLAGS() + 8 * 3 + 1 + 2 + 4 * N_CHUNKS();} // 54 for N=1, NC=6, NF=15
+
+function ALL_ONES() {return 0x1111111111111111;} 
+
+// state: [field; 2] = [step_num, pc]
+function STEP_NUM_IDX() {return 0;}
+function PC_IDX() {return 1;}
+
+// Utility functions
+template subarray(S, L, N) {
+    signal input in[N];
+    signal output out[L];
+
+    for (var i=S; i<S+L; i++) {
+        out[i-S] <== in[i];
+    }
+}
+
+template submatrix(S, R, C, N) {
+    signal input in[N];
+    signal output out[R][C];
+
+    for (var r=0; r<R; r++) {
+        for (var c=0; c<C; c++) {
+            out[r][c] <== in[S + r * R + c];
+        }
+    }
+}
+
+template if_else() {
+    signal input in[3]; 
+    signal output out; 
+
+    signal zero_for_a <== in[0];
+    signal a <== in[1];
+    signal b <== in[2];
+
+    signal _please_help_me_god <== (1-zero_for_a) * a;
+    out <== _please_help_me_god + (zero_for_a) * b;
+}
+
+template combine_chunks(N, L) {
+    signal input in[N];
+    signal output out;
+
+    signal combine[N];
+    for (var i=0; i<N; i++) {
+        if (i==0) {
+            combine[i] <== in[0];
+        } else {
+            combine[i] <== combine[i-1] + 2**(i*L) * in[i];
+        }
+    }
+    
+    out <== in[N-1];
+}
+
+// One CPU step of jolt 
+template JoltStep() {
+    /*  Input and Output states 
+    */
+    signal input input_state[2];
+    signal output output_state[2];
+
+    signal input witnesses[WIT_PER_STEP()];
+
+    /* The rest of this function has three broad components: 
+    1. Parse the witness to get the individual elements 
+    2. Write out the constraints
+    */
+
+    /*********************** 1. PARSE WITNESSES *******************************/
+    /* These Per-step witnesses
+    */
+
+    signal prog_read_avt[6][3];
+    signal op_flags_packed;
+    signal op_flag[N_FLAGS()]; // just advice, not read from code
+
+    /* Register reading witnesses 
+    */
+    signal rs1_read_avt[3];
```

Comment:
What are the expected values here? Is this a special (a,v,t) from RAM memory checking where a == rs1?